File: app/src/characteristics/installDockerImage.ts
import { createRequire } from 'module';

import { Logger } from '@utils/logger';
import { imagePull } from '@utils/docker';

enum InstallStatus {
	NOT_STARTED = 0,
	IN_PROGRESS = 1,
	COMPLETED = 2,
	ERROR = -1
}

export class InstallDockerImageCharacteristic
{
	/**
	 * Bleno instance
	 * @type any
	 */
	private Bleno: any = undefined;
	
	/**
	 * UUID of the characteristic
	 * @type string
	 */
	private characteristicUuid: string = '';
	
	/**
	 * Installation status
	 * @type InstallStatus
	 */
	private installStatus: InstallStatus = InstallStatus.NOT_STARTED;
	
	/**
	 * Create a new instance of Characteristic
	 */
	constructor(uuid: string)
	{
		const require = createRequire(import.meta.url);
		this.Bleno = require('bleno');
		this.characteristicUuid = uuid;
	}
	
	/**
	 * Create a new instance of InstallDockerImageCharacteristic
	 */
	public create()//: typeof Bleno.Characteristic 
	{
		if(this.Bleno === undefined)
			return null;
		
		return new this.Bleno.Characteristic({
			uuid: this.characteristicUuid,
			properties: ['read', 'write'],
			onReadRequest: this.onReadRequest.bind(this),
			onWriteRequest: this.onWriteRequest.bind(this),
		});
	}

	/**
	 * Called when the characteristic is read
	 * @param offset number
	 * @param callback (result: number, data: Buffer) => void
	 * @returns void
	 */
	public onReadRequest(offset: number, callback: (result: number, data: Buffer) => void)
	{
		let response;
		switch(this.installStatus)
		{
			case InstallStatus.NOT_STARTED: response = '0'; break;
			case InstallStatus.IN_PROGRESS: response = '1'; break;
			case InstallStatus.COMPLETED: response = '2'; break;
			case InstallStatus.ERROR: response = '-1'; break;
		}
		
		Logger.info(`Install Docker Image status: ${response}`);
		callback(this.Bleno.Characteristic.RESULT_SUCCESS, Buffer.from(response));
	}
	
	// Start the installation process with a write request
	public onWriteRequest(data: Buffer, offset: number, withoutResponse: boolean, callback: (result: number) => void)
	{
		if(this.installStatus === InstallStatus.IN_PROGRESS)
		{
			Logger.error('Installation already in progress');
			callback(this.Bleno.Characteristic.RESULT_UNLIKELY_ERROR);
			return;
		}
		
		// Start the installation process
		this.installStatus = InstallStatus.IN_PROGRESS;
		// Send a response to the client
		callback(this.Bleno.Characteristic.RESULT_SUCCESS);
		
		Logger.info('Starting Docker image installation');
		// Start the installation process
		imagePull()
		.then((status: boolean) =>
		{
			Logger.info(`Docker installation status: ${status}`);
			this.installStatus = status ? InstallStatus.COMPLETED : InstallStatus.ERROR;
		})
		.catch((error: any) =>
		{
			Logger.error(`Error while installing Docker image: ${error}`);
			this.installStatus = InstallStatus.ERROR;
		});
	}
}

File: ble/characteristics/install_docker_image.py
#!/usr/bin/env python3
import dbus
import dbus.service
import threading
from enum import Enum
from characteristics.base import BaseCharacteristic
from utils import logger
from utils.api import APIClient

class InstallStatus(Enum):
	NOT_STARTED = "0"
	IN_PROGRESS = "1"
	COMPLETED = "2"
	ERROR = "-1"

class InstallDockerImageCharacteristic(BaseCharacteristic):
	def __init__(self, bus, index, uuid):
		# This characteristic supports read write, and notify.
		flags = ['read', 'write', 'notify']
		super().__init__(bus, index, uuid, flags)
		self.service_path = '/org/bluez/example/service0'
		self.api_client = APIClient()
		# Initialize the installation status.
		self.install_status = InstallStatus.NOT_STARTED
		# Lock for thread-safety when modifying install_status.
		self.lock = threading.Lock()
		# Initialize the notifying flag.
		self.notifying = False

	@dbus.service.method("org.bluez.GattCharacteristic1", in_signature="a{sv}", out_signature="ay")
	def ReadValue(self, options):
		"""
		Called by a client to read the current installation status.
		Returns the state as a list of dbus.Byte values.
		"""
		with self.lock:
			current_status = self.install_status.value
		logger.info(f"InstallDockerImageCharacteristic: reading status '{current_status}'")
		return [dbus.Byte(b) for b in current_status.encode("utf-8")]

	@dbus.service.method("org.bluez.GattCharacteristic1", in_signature="aya{sv}", out_signature="")
	def WriteValue(self, value, options):
		"""
		Called by a client to trigger the installation process.
		Expects the string 'install' to start the process.
		"""
		action = bytes(value).decode("utf-8").strip().lower()
		logger.info(f"InstallDockerImageCharacteristic: WriteValue called with action '{action}'")
		if action == "install":
			with self.lock:
				# Check if an installation is already in progress.
				if self.install_status == InstallStatus.IN_PROGRESS:
					logger.error("InstallDockerImageCharacteristic: Installation already in progress")
					return
				self.install_status = InstallStatus.IN_PROGRESS
			# Start the installation in a separate thread.
			self._notify_clients()
			threading.Thread(target=self._install_docker_image, daemon=True).start()
		else:
			logger.error(f"InstallDockerImageCharacteristic: Unknown action '{action}'")
			with self.lock:
				self.install_status = InstallStatus.ERROR
			self._notify_clients()

	def _install_docker_image(self):
		"""
		Calls the API to install the Docker image and updates the installation status accordingly.
		The API call uses a timeout of 60 seconds.
		"""
		try:
			response = self.api_client.post("api/v1/install/docker-image", timeout=60)
			with self.lock:
				if response is not None:
					self.install_status = InstallStatus.COMPLETED
					logger.info("InstallDockerImageCharacteristic: Installation succeeded")
				else:
					self.install_status = InstallStatus.ERROR
					logger.error("InstallDockerImageCharacteristic: Installation failed (no response)")
			self._notify_clients()
		except Exception as e:
			with self.lock:
				self.install_status = InstallStatus.ERROR
			logger.error(f"InstallDockerImageCharacteristic: Error installing docker image: {e}")
			self._notify_clients()

	@dbus.service.method("org.bluez.GattCharacteristic1", in_signature="", out_signature="")
	def StartNotify(self):
		logger.info("[Docker] StartNotify")
		self.notifying = True

	@dbus.service.method("org.bluez.GattCharacteristic1", in_signature="", out_signature="")
	def StopNotify(self):
		logger.info("[Docker] StopNotify")
		self.notifying = False

	def _notify_clients(self):
		"""Emits a PropertiesChanged signal if a client is subscribed."""
		if not getattr(self, 'notifying', False):
			return
		with self.lock:
			val = [dbus.Byte(b) for b in self.install_status.value.encode()]
		self.PropertiesChanged(
			"org.bluez.GattCharacteristic1",
			{"Value": dbus.Array(val, signature="y")},
			[]
		)
